app/services/network_topology_service.rb

```diff
@@ -0,0 +1,118 @@
+            :FloatingIps    => nil,
+            :CloudTenant    => nil,
+            :SecurityGroups => nil
+          },
+          :NetworkRouter => {
+            :CloudNetwork => {
+              :FloatingIps => nil}
+          },
+        }
+      },
+    }
+
+    preloaded = @providers.includes(:cloud_subnets => [:cloud_network,
+                                                       :vms => [
+                                                         :floating_ips,
+                                                         :cloud_tenant,
+                                                         :security_groups],
+                                                       :network_router => [
+                                                         :cloud_network => [
+                                                           :floating_ips
+                                                         ]
+                                                       ]])
+    preloaded.each do |entity|
+      topo_items, links = build_recursive_topology(entity, entity_relationships[:NetworkManager], topo_items, links)
+    end
+
+    icons = {:CloudSubnet   => {:type => "glyph", :icon => "\uE909", :fontfamily => "PatternFlyIcons-webfont"}, # pficon-network
+             :NetworkRouter => {:type => "glyph", :icon => "\uE625", :fontfamily => "PatternFlyIcons-webfont"}, # pficon-route
+             :SecurityGroup => {:type => "glyph", :icon => "\uE903", :fontfamily => "PatternFlyIcons-webfont"}, # pficon-cloud-security
+             :FloatingIp    => {:type => "glyph", :icon => "\uF041", :fontfamily => "FontAwesome"},             # fa-map-marker
+             :CloudNetwork  => {:type => "glyph", :icon => "\uE62c", :fontfamily => "IcoMoon"},
+             :CloudTenant   => {:type => "glyph", :icon => "\uE904", :fontfamily => "PatternFlyIcons-webfont"}, # pficon-cloud-tenant
+             :Vm            => {:type => "glyph", :icon => "\uE90f", :fontfamily => "PatternFlyIcons-webfont"}, # pficon-virtual-machine
+             :Openstack     => {:type => "image", :icon => provider_icon(:Openstack)},
+             :Amazon        => {:type => "image", :icon => provider_icon(:Amazon)},
+             :Azure         => {:type => "image", :icon => provider_icon(:Azure)},
+             :Google        => {:type => "image", :icon => provider_icon(:Google)},
+    }
+
+    populate_topology(topo_items, links, build_kinds, icons)
+  end
+
+  def provider_icon(provider_type)
+    file_name = 'svg/vendor-' + provider_type.to_s.underscore.downcase + '.svg'
+    ActionController::Base.helpers.image_path(file_name)
+  end
+
+  def entity_display_type(entity)
+    if entity.kind_of?(ManageIQ::Providers::NetworkManager)
+      entity.class.short_token
+    else
+      name = entity.class.name.demodulize
+      if entity.kind_of?(Vm)
+        name.upcase # turn Vm to VM because it's an abbreviation
+      elsif ['Public', 'Private'].include?(name) && entity.kind_of?(CloudNetwork)
```

Comment:
@Ladas Any idea why are we treating CloudNetworks named "Public" or "Private" differently from any other cloud networks?

Reply:
I am not sure :-) What does the ```entity_type(entity)``` type returns?

Reply:
`entity.class.base_class.name.demodulize`